### PR TITLE
Stripe Payments

### PR DESCRIPTION
**Pull request by postables**
## :construction_worker: Purpose
* Allow purchasing credits with stripe

## :rocket: Changes
* Adds a stripe payment api call
* stripe-go dependency added

## :warning: Breaking Changes
* None

### CODE REVIEW

File: api/v2/routes_payment.go [hunk 1]
```diff
@@ -290,6 +293,57 @@ func (api *API) CreateDashPayment(c *gin.Context) {
...
+		Description: stripe.String("temporal credit purchase"),
+		// StatementDescriptor is what appears in their credit card billing report
+		StatementDescriptor: stripe.String("credit purchase"),
+		// email the receipt goes to
+		ReceiptEmail: stripe.String(forms["stripe_email"]),
+	}
+	// set the token generated by stripe
+	chargeParams.SetSource(forms["stripe_token"])
+	// metadata can be used for better Radar validation
+	chargeParams.AddMetadata("order_type", "temporal.credits")
```
Comment (bobheadxi) [inline api/v2/routes_payment.go]: can't these be set directly when we make `ChargeParams`?

File: api/v2/routes_payment.go [hunk 2]
```diff
@@ -290,6 +293,57 @@ func (api *API) CreateDashPayment(c *gin.Context) {
...
+	ch, err := charge.New(chargeParams)
+	if err != nil {
+		api.LogError(c, err, err.Error())(http.StatusBadRequest)
+		return
+	}
+	api.l.Infow("payment complete", "payment.method", "stripe", "user", username, "charge", ch)
+	// add credits to use
+	if _, err := api.um.AddCredits(username, valueInCentsFloat/100); err != nil {
+		api.LogError(c, err, "failed to grant credits")(http.StatusBadRequest)
```
Comment (bobheadxi) [inline api/v2/routes_payment.go]: is this really a bad request?

Comment (postables) [reply api/v2/routes_payment.go]: Good point I'll change to internal server error

File: api/v2/routes_payment.go (hunk 1 above)
Comment (postables) [reply api/v2/routes_payment.go]: no they have they have to be configured after instantiating `ChargeParams`

Comment (bobheadxi) [reply api/v2/routes_payment.go]: looks like you can (and this is safer too, protecting against invalid source):

```go
	// set the token generated by stripe
	source, err := stripe.SourceParamsFor(forms["stripe_token"])
	if err != nil {
		Fail(c, err)
		return
	}

	// initialize credit card charge parameters
	chargeParams := &stripe.ChargeParams{
		Amount:      stripe.Int64(valueInCentsInt),
		Currency:    stripe.String(string(stripe.CurrencyUSD)),
		Description: stripe.String("temporal credit purchase"),
		// StatementDescriptor is what appears in their credit card billing report
		StatementDescriptor: stripe.String("credit purchase"),
		// email the receipt goes to
		ReceiptEmail: stripe.String(forms["stripe_email"]),
		Source:       source,
		Params: stripe.Params{
			// metadata can be used for better Radar validation
			Metadata: map[string]string{"order_type": "temporal.credits"},
		},
	}
```

Comment (bobheadxi) [reply api/v2/routes_payment.go]: this gets rid of the setters, and you can now line the whole thing to make it easier to read:

```go
	// set the token generated by stripe
	source, err := stripe.SourceParamsFor(forms["stripe_token"])
	if err != nil {
		Fail(c, err)
		return
	}

	// create charge
	ch, err := charge.New(&stripe.ChargeParams{
		Amount:      stripe.Int64(valueInCentsInt),
		Currency:    stripe.String(string(stripe.CurrencyUSD)),
		Description: stripe.String("temporal credit purchase"),
		// StatementDescriptor is what appears in their credit card billing report
		StatementDescriptor: stripe.String("credit purchase"),
		// email the receipt goes to
		ReceiptEmail: stripe.String(forms["stripe_email"]),
		Source:       source,
		Params: stripe.Params{
			// metadata can be used for better Radar validation
			Metadata: map[string]string{"order_type": "temporal.credits"},
		},
	})
```

File: api/v2/routes_payment.go [hunk 3]
```diff
@@ -290,6 +293,57 @@ func (api *API) CreateDashPayment(c *gin.Context) {
...
+		return
+	}
+	valueInCentsInt, err := strconv.ParseInt(forms["value_in_cents"], 10, 64)
+	if err != nil {
+		Fail(c, err)
+		return
+	}
+	valueInCentsFloat, err := strconv.ParseFloat(forms["value_in_cents"], 64)
+	if err != nil {
+		Fail(c, err)
```
Comment (bobheadxi) [inline api/v2/routes_payment.go]:
```suggestion
		Fail(c, err)
		return
```

File: api/v2/routes_payment.go (hunk 1 above)
Comment (bobheadxi) [reply api/v2/routes_payment.go]: source: https://sourcegraph.com/github.com/RTradeLtd/Temporal@payments/stripe/-/blob/vendor/github.com/stripe/stripe-go/charge.go#L86:28

File: api/v2/routes_payment.go (hunk 3 above)
Comment (postables) [reply api/v2/routes_payment.go]: :thinking:  looks the same

File: api/v2/routes_payment.go (hunk 1 above)
Comment (postables) [reply api/v2/routes_payment.go]: slick

File: api/v2/routes_payment.go (hunk 3 above)
Comment (bobheadxi) [reply api/v2/routes_payment.go]: you forgot a `return`

Comment (postables) [reply api/v2/routes_payment.go]: ah kk

File: api/v2/routes_payment.go [hunk 4]
```diff
@@ -7,8 +7,11 @@ import (
 	"net/http"
 	"strconv"
 
+	"github.com/stripe/stripe-go/charge"
+
 	"github.com/RTradeLtd/Temporal/eh"
 	"github.com/RTradeLtd/gorm"
+	"github.com/stripe/stripe-go"
```
Comment (bobheadxi) [inline api/v2/routes_payment.go]: nitpick: maybe group the imports together?

**Pull request merged**